### PR TITLE
[Clang] warn on discarded [[nodiscard]] function results after casting in C

### LINKED ISSUE #104391: [clang][c23] macro + non-void cast incorrectly suppresses "nodiscard" warning
clang-18 compiles the following code warning-free at `-Weverything`, but should warn because the result of `f()` is essentially discarded (aside from the cast, but it would be to void if the result was supposed to be ignorable):
```c
[[nodiscard]] int f();

void g() {
#define M (unsigned int)f()
  M;
}
```
https://godbolt.org/z/9sdcY3eof

It does correctly warn when compiling as C++ (even with the C-style cast).

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #104391

### CODE REVIEW

**Comment by cor3ntin** [comment]
can you add a test for void? we should make sure this still warns!

**Comment by a-tarasyuk** [comment]
@cor3ntin Do you mean casting to `void` or declaring a `void` function?

**Comment by cor3ntin** [comment]
Casting to void

File: clang/test/Sema/c2x-nodiscard.c [hunk 1]
```diff
@@ -54,3 +54,9 @@ void test_missiles(void) {
   launch_missiles();
 }
 
+[[nodiscard]] int f3();
+
+void f4() {
```
Comment (cor3ntin) [inline clang/test/Sema/c2x-nodiscard.c:59]:
```suggestion
void GH104391(void) {
```

Comment (a-tarasyuk) [reply clang/test/Sema/c2x-nodiscard.c]: @cor3ntin I've updated the test. Thanks for the review.

**Review by cor3ntin — APPROVED** [review]

File: clang/test/Sema/c2x-nodiscard.c (hunk 1 above)
Comment (a-tarasyuk) [reply clang/test/Sema/c2x-nodiscard.c]: @cor3ntin could you merge it once the CI passes? Thanks

**Pull request merged**